File: app/src/main/java/com/whaleshark/basicarch/data/SearchRepository.java
package com.whaleshark.basicarch.data;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import com.whaleshark.basicarch.api.FoodPuppyService;
import com.whaleshark.basicarch.model.Recipe;
import com.whaleshark.basicarch.model.RecipeSearchResult;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * @author stipton
 */

public class SearchRepository {

    private final FoodPuppyService foodPuppyService = FoodPuppyService.Factory.create();

    public LiveData<List<Recipe>> searchRecipes(String query) {
        MutableLiveData<List<Recipe>> result = new MutableLiveData<>();

        foodPuppyService.searchRecipes(query).enqueue(new Callback<RecipeSearchResult>() {
            @Override
            public void onResponse(Call<RecipeSearchResult> call, Response<RecipeSearchResult> response) {
                result.postValue(response.body().results);
            }

            @Override
            public void onFailure(Call<RecipeSearchResult> call, Throwable t) {
                result.postValue(null);
            }
        });

        return result;
    }
}

File: app/src/main/java/com/whaleshark/basicarch/data/SearchRepository.kt
package com.whaleshark.basicarch.data

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData

import com.whaleshark.basicarch.api.FoodPuppyService
import com.whaleshark.basicarch.model.Recipe
import com.whaleshark.basicarch.model.RecipeSearchResult

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * @author stipton
 */

class SearchRepository {

    private val foodPuppyService = FoodPuppyService.Factory.create()

    fun searchRecipes(query: String): LiveData<List<Recipe>> {
        val result = MutableLiveData<List<Recipe>>()

        foodPuppyService.searchRecipes(query).enqueue(object : Callback<RecipeSearchResult> {
            override fun onResponse(call: Call<RecipeSearchResult>, response: Response<RecipeSearchResult>) {
                result.postValue(response.body()!!.results)
            }

            override fun onFailure(call: Call<RecipeSearchResult>, t: Throwable) {
                result.postValue(null)
            }
        })

        return result
    }
}
